tests/rate_limited.rs

```diff
@@ -0,0 +1,325 @@
+            "hostnames": ["*.toystore.com", "example.com"],
+            "gateway_actions": [
+            {
+                "rules": [
+                {
+                    "paths": ["/admin/toy"],
+                    "hosts": ["cars.toystore.com"],
+                    "methods": ["POST"]
+                }],
+                "configurations": [
+                {
+                    "actions": [
+                    {
+                        "generic_key": {
+                            "descriptor_key": "admin",
+                            "descriptor_value": "1"
+                        }
+                    }
+                    ]
+                }
+                ]
+            }
+            ]
+        }
+        ]
+    }"#;
+
+    module
+        .call_proxy_on_context_create(root_context, 0)
+        .expect_log(Some(LogLevel::Info), Some("set_root_context #1"))
+        .execute_and_expect(ReturnType::None)
+        .unwrap();
+    module
+        .call_proxy_on_configure(root_context, 0)
+        .expect_log(Some(LogLevel::Info), Some("on_configure #1"))
+        .expect_get_buffer_bytes(Some(BufferType::PluginConfiguration))
+        .returning(Some(cfg.as_bytes()))
+        .expect_log(Some(LogLevel::Info), None)
+        .execute_and_expect(ReturnType::Bool(true))
+        .unwrap();
+
+    let http_context = 2;
+    module
+        .call_proxy_on_context_create(http_context, root_context)
+        .expect_log(Some(LogLevel::Info), Some("create_http_context #2"))
+        .execute_and_expect(ReturnType::None)
+        .unwrap();
+
+    module
+        .call_proxy_on_request_headers(http_context, 0, false)
+        .expect_log(Some(LogLevel::Info), Some("on_http_request_headers #2"))
+        .expect_get_header_map_value(Some(MapType::HttpRequestHeaders), Some(":authority"))
+        .returning(Some("cars.toystore.com"))
+        .expect_get_header_map_value(Some(MapType::HttpRequestHeaders), Some(":path"))
+        .returning(Some("/admin/toy"))
+        .expect_get_header_map_value(Some(MapType::HttpRequestHeaders), Some(":method"))
+        .returning(Some("POST"))
+        .expect_get_header_map_value(Some(MapType::HttpRequestHeaders), Some(":authority"))
+        .returning(Some("cars.toystore.com"))
+        .expect_grpc_call(
+            Some("limitador-cluster"),
+            Some("envoy.service.ratelimit.v3.RateLimitService"),
+            Some("ShouldRateLimit"),
+            Some(&[0, 0, 0, 0]),
+            Some(&[
+                10, 10, 82, 76, 83, 45, 100, 111, 109, 97, 105, 110, 18, 12, 10, 10, 10, 5, 97,
+                100, 109, 105, 110, 18, 1, 49, 24, 1,
+            ]),
+            Some(5000),
+        )
+        .returning(Some(42))
+        .expect_log(
+            Some(LogLevel::Info),
+            Some("Initiated gRPC call (id# 42) to Limitador"),
+        )
+        .execute_and_expect(ReturnType::Action(Action::Pause))
+        .unwrap();
+
+    let grpc_response: [u8; 45] = [
```

Comment:
Why not encode like this? 

```rust
        let mut resp = RateLimitResponse::new();
        resp.overall_code = RateLimitResponse_Code::OK;
        resp.response_headers_to_add
            .push(header("test", "some value"));
        resp.response_headers_to_add
            .push(header("other", "header value"));
        let buffer = resp.write_to_bytes().unwrap();
        let grpc_response: [u8; 45] = buffer.as_slice();
```

I am not very familiar at reading bytes

Reply:
We can't because of the lib type (`cdylib`), the tests in `tests` can't reference the code under test for… reasons. I couldn't get it to target another type, because of unresolved link time deps. That code, the envoy types, are part of the wasm-shim code and not a dependency, as they are generated using `protoc`.
As I mentioned in the sync, we could generate these twice, once as we do now, and another time in the `tests` dir, so we could use them there. But, just like for the PluginConfiguration, which uses a JSON format, I decided to use the serialized form directly, i.e. protobuf… hence the test in `lib.rs` that asserts that this is a valid byte sequence.